From c23ab3cabeb6fa3ffedc4c29350d4ce49faf1faf Mon Sep 17 00:00:00 2001
From: Jim Somerville <Jim.Somerville@windriver.com>
Date: Fri, 10 Apr 2026 13:21:28 -0300
Subject: [PATCH] STX: Suspend/Resume for VMs with PCIPT+Virtio

When we have a Passthrough or a SR-IOV device, the plug/unplug of these
devices causes the guest memory map to change. After VHOST_SET_MEM_TABLE,
send VHOST_SET_VRING_ADDR to force vring mapping recalculation based on
the new memory map.

Signed-off-by: Jim Somerville <Jim.Somerville@windriver.com>
[ Trimmed the shortlog ]
Signed-off-by: Rafael Falcao <Rafael.VieiraFalcao@windriver.com>
[Rebased original changes to the qemu 7.2 version]
Signed-off-by: david.liu <david.liu@windriver.com>
[Rebased to qemu 10.0.8 for Trixie]
Signed-off-by: cbohlhal <Caique.BohlhalterdeSouza@windriver.com>
---
 hw/virtio/vhost.c | 26 +++++++++++++++++++++++++-
 1 file changed, 25 insertions(+), 1 deletion(-)

diff --git a/hw/virtio/vhost.c b/hw/virtio/vhost.c
index 971460eb2a..d3bb692eb6 100644
--- a/hw/virtio/vhost.c
+++ b/hw/virtio/vhost.c
@@ -81,6 +81,10 @@ unsigned int vhost_get_free_memslots(void)
     return free;
 }
 
+static int vhost_virtqueue_set_addr(struct vhost_dev *dev,
+                                    struct vhost_virtqueue *vq,
+                                    unsigned idx, bool enable_log);
+
 static void vhost_dev_sync_region(struct vhost_dev *dev,
                                   MemoryRegionSection *section,
                                   uint64_t mfirst, uint64_t mlast,
@@ -615,6 +619,21 @@ static void vhost_begin(MemoryListener *listener)
     dev->n_tmp_sections = 0;
 }
 
+static void vhost_update_backend_ring_mappings(struct vhost_dev *dev)
+{
+    int i,r;
+    
+    if(dev->vhost_ops->backend_type != VHOST_BACKEND_TYPE_USER) {
+        return;
+    }
+
+    for (i = 0; i < dev->nvqs; ++i) {
+        r = vhost_virtqueue_set_addr(dev, dev->vqs + i, i, dev->log_enabled);
+        assert(r >= 0);
+    }
+    return;
+}
+
 static void vhost_commit(MemoryListener *listener)
 {
     struct vhost_dev *dev = container_of(listener, struct vhost_dev,
@@ -691,7 +710,7 @@ static void vhost_commit(MemoryListener *listener)
         if (r < 0) {
             VHOST_OPS_DEBUG(r, "vhost_set_mem_table failed");
         }
-        goto out;
+        goto vring_mapping;
     }
     log_size = vhost_get_log_size(dev);
     /* We allocate an extra 4K bytes to log,
@@ -710,6 +729,11 @@ static void vhost_commit(MemoryListener *listener)
         vhost_dev_log_resize(dev, log_size);
     }
 
+vring_mapping:
+/* For vhost-user backend, update the vring mappings after we sent a new
+* guest memory map. */
+    vhost_update_backend_ring_mappings(dev);	
+
 out:
     /* Deref the old list of sections, this must happen _after_ the
      * vhost_set_mem_table to ensure the client isn't still using the
-- 
2.34.1

